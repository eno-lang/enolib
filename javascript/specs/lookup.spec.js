import { lookup } from '..';

// TODO: Error case lookups (indices -1/length+1 etc. and same for line/column)
//       (all implementations)

const input = `
> comment

field: value

list:
- item

> comment
- item

fieldset:
entry = value

> comment
entry = value

# section

> comment
-- empty
-- empty

-- multiline
value
value
-- multiline

## subsection
`.trim();

const SNIPPET_PADDING_WIDTH = 3
const SNIPPET_PADDING = '▓'.repeat(SNIPPET_PADDING_WIDTH);

// The + ' ' before closing SNIPPET_PADDING represents the last cursor
// (there is always one cursor index more than there are chars in a string)
const snippetInput = SNIPPET_PADDING + input.replace(/\n/g, '⏎').replace(/\t/g, '⇥').replace(/ /g, '␣') + ' ' + SNIPPET_PADDING;

const snippet = index => {
    index += SNIPPET_PADDING_WIDTH;
    
    return snippetInput.substring(index - SNIPPET_PADDING_WIDTH, index) +
           '  ' + snippetInput.charAt(index) + '  ' +
           snippetInput.substring(index + 1, index + SNIPPET_PADDING_WIDTH + 1);
};

describe('lookup', () => {
    let column = 0;
    let line = 0;
    
    let summary = '\nINDEX  SNIPPET            KEY                  RANGE\n\n';
    
    for (let index = 0; index <= input.length; index++) {
        const indexLookup = lookup({ index }, input);
        const lineColumnLookup = lookup({ line, column }, input);
        
        if (indexLookup.range !== lineColumnLookup.range)
            throw new Error(`Lookup by index produced a different range (${indexLookup.range}) than by line/column (${lineColumnLookup.range})`);
        
        if (indexLookup.element.stringKey() !== lineColumnLookup.element.stringKey())
            throw new Error(`Lookup by index produced a different key (${indexLookup.element.stringKey()}) than by line/column (${lineColumnLookup.element.stringKey()})`);
        
        const { element, range } = indexLookup;
        const key = element.stringKey() === null ? 'document' : element.stringKey();
        summary += `${index.toString().padEnd(5)}  ${snippet(index).padStart(9)}   =>   ${key.padEnd(20)} ${range}\n`;
        
        if (input.charAt(index) === '\n') {
            line++;
            column = 0;
        } else {
            column++;
        }
    }
    
    it(`produces the expected summary`, () => {
        expect(summary).toMatchSnapshot();
    });
});
