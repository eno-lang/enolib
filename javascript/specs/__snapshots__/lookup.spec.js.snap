// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lookup produces the expected summary 1`] = `
"
INDEX  SNIPPET            KEY                  RANGE

0      ▓▓▓  >  ␣co   =>   document             commentOperator
1      ▓▓>  ␣  com   =>   document             commentOperator
2      ▓>␣  c  omm   =>   document             comment
3      >␣c  o  mme   =>   document             comment
4      ␣co  m  men   =>   document             comment
5      com  m  ent   =>   document             comment
6      omm  e  nt⏎   =>   document             comment
7      mme  n  t⏎⏎   =>   document             comment
8      men  t  ⏎⏎f   =>   document             comment
9      ent  ⏎  ⏎fi   =>   document             comment
10     nt⏎  ⏎  fie   =>   document             null
11     t⏎⏎  f  iel   =>   field                key
12     ⏎⏎f  i  eld   =>   field                key
13     ⏎fi  e  ld:   =>   field                key
14     fie  l  d:␣   =>   field                key
15     iel  d  :␣v   =>   field                key
16     eld  :  ␣va   =>   field                fieldOperator
17     ld:  ␣  val   =>   field                fieldOperator
18     d:␣  v  alu   =>   field                value
19     :␣v  a  lue   =>   field                value
20     ␣va  l  ue⏎   =>   field                value
21     val  u  e⏎⏎   =>   field                value
22     alu  e  ⏎⏎f   =>   field                value
23     lue  ⏎  ⏎fi   =>   field                value
24     ue⏎  ⏎  fie   =>   document             null
25     e⏎⏎  f  iel   =>   field_with_items     key
26     ⏎⏎f  i  eld   =>   field_with_items     key
27     ⏎fi  e  ld_   =>   field_with_items     key
28     fie  l  d_w   =>   field_with_items     key
29     iel  d  _wi   =>   field_with_items     key
30     eld  _  wit   =>   field_with_items     key
31     ld_  w  ith   =>   field_with_items     key
32     d_w  i  th_   =>   field_with_items     key
33     _wi  t  h_i   =>   field_with_items     key
34     wit  h  _it   =>   field_with_items     key
35     ith  _  ite   =>   field_with_items     key
36     th_  i  tem   =>   field_with_items     key
37     h_i  t  ems   =>   field_with_items     key
38     _it  e  ms:   =>   field_with_items     key
39     ite  m  s:⏎   =>   field_with_items     key
40     tem  s  :⏎-   =>   field_with_items     key
41     ems  :  ⏎-␣   =>   field_with_items     fieldOperator
42     ms:  ⏎  -␣i   =>   field_with_items     fieldOperator
43     s:⏎  -  ␣it   =>   field_with_items     itemOperator
44     :⏎-  ␣  ite   =>   field_with_items     itemOperator
45     ⏎-␣  i  tem   =>   field_with_items     value
46     -␣i  t  em⏎   =>   field_with_items     value
47     ␣it  e  m⏎⏎   =>   field_with_items     value
48     ite  m  ⏎⏎>   =>   field_with_items     value
49     tem  ⏎  ⏎>␣   =>   field_with_items     value
50     em⏎  ⏎  >␣c   =>   field_with_items     null
51     m⏎⏎  >  ␣co   =>   field_with_items     commentOperator
52     ⏎⏎>  ␣  com   =>   field_with_items     commentOperator
53     ⏎>␣  c  omm   =>   field_with_items     comment
54     >␣c  o  mme   =>   field_with_items     comment
55     ␣co  m  men   =>   field_with_items     comment
56     com  m  ent   =>   field_with_items     comment
57     omm  e  nt⏎   =>   field_with_items     comment
58     mme  n  t⏎-   =>   field_with_items     comment
59     men  t  ⏎-␣   =>   field_with_items     comment
60     ent  ⏎  -␣i   =>   field_with_items     comment
61     nt⏎  -  ␣it   =>   field_with_items     itemOperator
62     t⏎-  ␣  ite   =>   field_with_items     itemOperator
63     ⏎-␣  i  tem   =>   field_with_items     value
64     -␣i  t  em⏎   =>   field_with_items     value
65     ␣it  e  m⏎⏎   =>   field_with_items     value
66     ite  m  ⏎⏎f   =>   field_with_items     value
67     tem  ⏎  ⏎fi   =>   field_with_items     value
68     em⏎  ⏎  fie   =>   document             null
69     m⏎⏎  f  iel   =>   field_with_attributes key
70     ⏎⏎f  i  eld   =>   field_with_attributes key
71     ⏎fi  e  ld_   =>   field_with_attributes key
72     fie  l  d_w   =>   field_with_attributes key
73     iel  d  _wi   =>   field_with_attributes key
74     eld  _  wit   =>   field_with_attributes key
75     ld_  w  ith   =>   field_with_attributes key
76     d_w  i  th_   =>   field_with_attributes key
77     _wi  t  h_a   =>   field_with_attributes key
78     wit  h  _at   =>   field_with_attributes key
79     ith  _  att   =>   field_with_attributes key
80     th_  a  ttr   =>   field_with_attributes key
81     h_a  t  tri   =>   field_with_attributes key
82     _at  t  rib   =>   field_with_attributes key
83     att  r  ibu   =>   field_with_attributes key
84     ttr  i  but   =>   field_with_attributes key
85     tri  b  ute   =>   field_with_attributes key
86     rib  u  tes   =>   field_with_attributes key
87     ibu  t  es:   =>   field_with_attributes key
88     but  e  s:⏎   =>   field_with_attributes key
89     ute  s  :⏎a   =>   field_with_attributes key
90     tes  :  ⏎at   =>   field_with_attributes fieldOperator
91     es:  ⏎  att   =>   field_with_attributes fieldOperator
92     s:⏎  a  ttr   =>   attribute            key
93     :⏎a  t  tri   =>   attribute            key
94     ⏎at  t  rib   =>   attribute            key
95     att  r  ibu   =>   attribute            key
96     ttr  i  but   =>   attribute            key
97     tri  b  ute   =>   attribute            key
98     rib  u  te␣   =>   attribute            key
99     ibu  t  e␣=   =>   attribute            key
100    but  e  ␣=␣   =>   attribute            key
101    ute  ␣  =␣v   =>   attribute            key
102    te␣  =  ␣va   =>   attribute            attributeOperator
103    e␣=  ␣  val   =>   attribute            attributeOperator
104    ␣=␣  v  alu   =>   attribute            value
105    =␣v  a  lue   =>   attribute            value
106    ␣va  l  ue⏎   =>   attribute            value
107    val  u  e⏎⏎   =>   attribute            value
108    alu  e  ⏎⏎>   =>   attribute            value
109    lue  ⏎  ⏎>␣   =>   attribute            value
110    ue⏎  ⏎  >␣c   =>   field_with_attributes null
111    e⏎⏎  >  ␣co   =>   attribute            commentOperator
112    ⏎⏎>  ␣  com   =>   attribute            commentOperator
113    ⏎>␣  c  omm   =>   attribute            comment
114    >␣c  o  mme   =>   attribute            comment
115    ␣co  m  men   =>   attribute            comment
116    com  m  ent   =>   attribute            comment
117    omm  e  nt⏎   =>   attribute            comment
118    mme  n  t⏎a   =>   attribute            comment
119    men  t  ⏎at   =>   attribute            comment
120    ent  ⏎  att   =>   attribute            comment
121    nt⏎  a  ttr   =>   attribute            key
122    t⏎a  t  tri   =>   attribute            key
123    ⏎at  t  rib   =>   attribute            key
124    att  r  ibu   =>   attribute            key
125    ttr  i  but   =>   attribute            key
126    tri  b  ute   =>   attribute            key
127    rib  u  te␣   =>   attribute            key
128    ibu  t  e␣=   =>   attribute            key
129    but  e  ␣=␣   =>   attribute            key
130    ute  ␣  =␣v   =>   attribute            key
131    te␣  =  ␣va   =>   attribute            attributeOperator
132    e␣=  ␣  val   =>   attribute            attributeOperator
133    ␣=␣  v  alu   =>   attribute            value
134    =␣v  a  lue   =>   attribute            value
135    ␣va  l  ue⏎   =>   attribute            value
136    val  u  e⏎⏎   =>   attribute            value
137    alu  e  ⏎⏎#   =>   attribute            value
138    lue  ⏎  ⏎#␣   =>   attribute            value
139    ue⏎  ⏎  #␣s   =>   document             null
140    e⏎⏎  #  ␣se   =>   section              sectionOperator
141    ⏎⏎#  ␣  sec   =>   section              sectionOperator
142    ⏎#␣  s  ect   =>   section              key
143    #␣s  e  cti   =>   section              key
144    ␣se  c  tio   =>   section              key
145    sec  t  ion   =>   section              key
146    ect  i  on⏎   =>   section              key
147    cti  o  n⏎⏎   =>   section              key
148    tio  n  ⏎⏎>   =>   section              key
149    ion  ⏎  ⏎>␣   =>   section              key
150    on⏎  ⏎  >␣c   =>   section              null
151    n⏎⏎  >  ␣co   =>   embed                commentOperator
152    ⏎⏎>  ␣  com   =>   embed                commentOperator
153    ⏎>␣  c  omm   =>   embed                comment
154    >␣c  o  mme   =>   embed                comment
155    ␣co  m  men   =>   embed                comment
156    com  m  ent   =>   embed                comment
157    omm  e  nt⏎   =>   embed                comment
158    mme  n  t⏎-   =>   embed                comment
159    men  t  ⏎--   =>   embed                comment
160    ent  ⏎  --␣   =>   embed                comment
161    nt⏎  -  -␣e   =>   embed                embedOperator
162    t⏎-  -  ␣em   =>   embed                embedOperator
163    ⏎--  ␣  emb   =>   embed                embedOperator
164    --␣  e  mbe   =>   embed                key
165    -␣e  m  bed   =>   embed                key
166    ␣em  b  ed⏎   =>   embed                key
167    emb  e  d⏎-   =>   embed                key
168    mbe  d  ⏎--   =>   embed                key
169    bed  ⏎  --␣   =>   embed                key
170    ed⏎  -  -␣e   =>   embed                embedOperator
171    d⏎-  -  ␣em   =>   embed                embedOperator
172    ⏎--  ␣  emb   =>   embed                embedOperator
173    --␣  e  mbe   =>   embed                key
174    -␣e  m  bed   =>   embed                key
175    ␣em  b  ed⏎   =>   embed                key
176    emb  e  d⏎⏎   =>   embed                key
177    mbe  d  ⏎⏎-   =>   embed                key
178    bed  ⏎  ⏎--   =>   embed                key
179    ed⏎  ⏎  --␣   =>   section              null
180    d⏎⏎  -  -␣e   =>   embed                embedOperator
181    ⏎⏎-  -  ␣em   =>   embed                embedOperator
182    ⏎--  ␣  emb   =>   embed                embedOperator
183    --␣  e  mbe   =>   embed                key
184    -␣e  m  bed   =>   embed                key
185    ␣em  b  ed⏎   =>   embed                key
186    emb  e  d⏎v   =>   embed                key
187    mbe  d  ⏎va   =>   embed                key
188    bed  ⏎  val   =>   embed                key
189    ed⏎  v  alu   =>   embed                value
190    d⏎v  a  lue   =>   embed                value
191    ⏎va  l  ue⏎   =>   embed                value
192    val  u  e⏎v   =>   embed                value
193    alu  e  ⏎va   =>   embed                value
194    lue  ⏎  val   =>   embed                value
195    ue⏎  v  alu   =>   embed                value
196    e⏎v  a  lue   =>   embed                value
197    ⏎va  l  ue⏎   =>   embed                value
198    val  u  e⏎-   =>   embed                value
199    alu  e  ⏎--   =>   embed                value
200    lue  ⏎  --␣   =>   embed                value
201    ue⏎  -  -␣e   =>   embed                embedOperator
202    e⏎-  -  ␣em   =>   embed                embedOperator
203    ⏎--  ␣  emb   =>   embed                embedOperator
204    --␣  e  mbe   =>   embed                key
205    -␣e  m  bed   =>   embed                key
206    ␣em  b  ed⏎   =>   embed                key
207    emb  e  d⏎⏎   =>   embed                key
208    mbe  d  ⏎⏎#   =>   embed                key
209    bed  ⏎  ⏎##   =>   embed                key
210    ed⏎  ⏎  ##␣   =>   section              null
211    d⏎⏎  #  #␣s   =>   subsection           sectionOperator
212    ⏎⏎#  #  ␣su   =>   subsection           sectionOperator
213    ⏎##  ␣  sub   =>   subsection           sectionOperator
214    ##␣  s  ubs   =>   subsection           key
215    #␣s  u  bse   =>   subsection           key
216    ␣su  b  sec   =>   subsection           key
217    sub  s  ect   =>   subsection           key
218    ubs  e  cti   =>   subsection           key
219    bse  c  tio   =>   subsection           key
220    sec  t  ion   =>   subsection           key
221    ect  i  on    =>   subsection           key
222    cti  o  n ▓   =>   subsection           key
223    tio  n   ▓▓   =>   subsection           key
224    ion     ▓▓▓   =>   subsection           key
"
`;
